pkg/client/cache/reflector.go

```diff
@@ -306,13 +306,34 @@ func (r *Reflector) ListAndWatch(stopCh <-chan struct{}) error {
 	}
 	r.setLastSyncResourceVersion(resourceVersion)
 
+	resyncerrc := make(chan error, 1)
+	go func() {
+		for {
+			select {
+			case <-resyncCh:
+			case <-stopCh:
+				return
+			}
+			if r.canForceResyncNow() {
+				glog.V(4).Infof("%s: next resync planned for %#v, forcing now", r.name, r.nextResync)
+				if err := r.store.Resync(); err != nil {
+					resyncerrc <- err
+					return
+				}
+				cleanup()
+				resyncCh, cleanup = r.resyncChan()
+			}
+		}
+	}()
+
 	for {
-		options := api.ListOptions{
+		options = api.ListOptions{
 			ResourceVersion: resourceVersion,
 			// We want to avoid situations when resyncing is breaking the TCP connection
 			// - see comment for 'timeoutForWatch()' for more details.
 			TimeoutSeconds: r.timeoutForWatch(),
 		}
+
```

Comment:
As I wrote in the PR comment - with this approach timeoutForWatch can be signiifcantly simplied.
So it should be just:

     time.Duration(float64(minWatchTimeout) * (rand.Float64() + 1.0))

and we don't even need a function for it.
It was necessary when relist was strictly connected to resync (i.e. we couldnt' do resync with watch opened).